src/back/glsl450.rs

```diff
@@ -0,0 +1,201 @@
+use crate::{
+    back::glsl_common::{Error, StatementBuilder},
+    Binding, BuiltIn, FastHashMap, Module, TypeInner,
+};
+use std::{
+    fmt::{self, Write as FmtWrite},
+    io::Write,
+};
+
+pub fn write(module: &Module, out: &mut impl Write) -> Result<(), Error> {
+    writeln!(out, "#version 450 core")?;
+
+    let mut counter = 0;
+    let mut names = FastHashMap::default();
+
+    let mut namer = |name: Option<&String>| {
+        if let Some(name) = name {
+            names.insert(name.clone(), ());
+            name.clone()
+        } else {
+            counter += 1;
+            while names.get(&format!("_{}", counter)).is_some() {
+                counter += 1;
+            }
+            format!("_{}", counter)
+        }
+    };
+
+    let mut structs = FastHashMap::default();
+
+    // Do a first pass to collect names
+    for (handle, ty) in module.types.iter() {
+        match &ty.inner {
+            TypeInner::Struct { members } => {
+                let name = namer(ty.name.as_ref());
+                let mut idxs = Vec::new();
+
+                for member in members {
+                    idxs.push(namer(member.name.as_ref()))
+                }
+
+                structs.insert(handle, (name, idxs));
+            }
+            _ => continue,
+        }
+    }
+
+    // Do a second pass to build the structs
+    // TODO: glsl is order dependent so we need to build structs in order
+    for (handle, ty) in module.types.iter() {
+        match &ty.inner {
```

Comment:
please don't match on a reference. Rust makes smarts to let you do this in 2018 edition, but I'd really like our code base to not rely on these smarts. It's easier to reason about borrowing vs copying without that.